### ISSUE TITLE
Wrong highlight when using inline dtd in xml document

### ISSUE THREAD

**Issue by DerFrZocker** [pygments/pygments#2692]
When trying to highlight following xml, the `]>` at the end of `DOCTYPE` gets highlighted as normal text.
Possible because, in dtd the `ELEMENT` does not have a closing `/`.
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE hello [<!ELEMENT hello (#PCDATA)>]>
<hello>Hello Word!</hello>
```
Using:
```python
from pygments import highlight
from pygments.lexers import XmlLexer
from pygments.formatters import HtmlFormatter

code = '<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE hello [<!ELEMENT hello (#PCDATA)>]><hello>Hello Word!</hello>'
print(hightlight(code, XmlLexer(), HtmlFormatter()))
```

Results in:
![grafik](https://github.com/pygments/pygments/assets/39651319/f7d379f5-32e8-4e3a-aeae-ef88b12526d2)

https://pygments.org/demo/?lexer=xml&style=default&formatter=html&code=%3C%3Fxml+version%3D%221.0%22+encoding%3D%22UTF-8%22%3F%3E%0A%3C%21DOCTYPE+hello+%5B%3C%21ELEMENT+hello+%28%23PCDATA%29%3E%5D%3E%0A%3Chello%3EHello+Word%21%3C%2Fhello%3E

**Comment by jofish920** [comment]
Using the raw formatter makes this even clearer:

    Token.Comment.Preproc   '<?xml version="1.0" encoding="UTF-8"?>'
    Token.Text.Whitespace   '\n'
    Token.Comment.Preproc   '<!DOCTYPE hello [<!ELEMENT hello (#PCDATA)>'
    Token.Text      ']>'
    Token.Text.Whitespace   '\n'
    Token.Name.Tag  '<hello'
    ...

Which leads to the following on line 219 in the current version of `pygments.lexers.html.XmlLexer`:

    ('<![^>]*>', Comment.Preproc),
    
This will treat the first '>' as the end of the directive, giving the observed result.

There is a nice DTD lexer included in the same file, and it would be interesting to see if it is feasible to get the two lexers working together.

**Comment by birkenfeld** [comment]
If the DTD lexer can handle that content, you can use `using(DtdLexer)` to invoke it for a matching group.

**Comment by jofish920** [comment]
Thanks!  I'll look into it.

**Comment by jofish920** [comment]
Had a glance at `using`, and it seems to rely on the whole expression being processed as part of another language to be reconised by a regex first.  This is slightly problematic, as the DTD parser needs to be able to handle nested "[...]" and "<...>", which go beyond the capacity of regular expressions into the domain of context free languages (it would be easy if ']' or '>' could not be part of a DTD, or if there were not contexts like strings in which unpaired delimiters could legitamately occur).  

The problem is by no means impossible to solve, and I can already think of a couple of solutions, it's just finding the one that is the least ugly for the moment.  One option would be to set things up using a delegating lexer approach, copying just enough of the functionality of the DTD lexer to the Xml lexer to handle matching square brackets (taking into account things like brackets embedded in strings).  That is probably the cleanest approach.

Anyway, I'll get back to this when I have a bit more time.

**Comment by jofish920** [comment]
The commit referred to above is very much not ready for inclusion into anything; it was a result of me messing around with the code while avoiding work on something else.    I think that it provides a valid approach to solving the problem, and it could easily be turned into a proper fix.  I will work on it later, if nobody else does :)

**Comment by birkenfeld** [comment]
Yes, DelegatingLexer is the next step of "escalation" if `using` doesn't do the trick :)

**Comment by jofish920** [comment]
I finally looked up the examples that I recalled showing [...] in a DTD to put them in a test.  I found several in the DTDs for various versions of HTML.  They all occur in strings, comment sections, either standalone or in definitions (--...--), or in constructs like "<![CDATA[ ... ]]>".  My tentative solution involving a simple context-free sublanguage may be neither necessary nor (in its current form) sufficient.  I don't anticipate many cases of nested DOCTYPE declarations, which is what I think I was considering when I decided that inline DTDs required at least basic context free parsing.  Anyway, back to what I shold be working on.

**Comment by jofish920** [comment]
If anyone has any comment on PR #2810, or any idea on why `regexlint` might be raising an exception when processing the modified lexers, I'd be glad for any feedback.  I have considered moving this out of draft mode, but really don't want to while it is still producing errors on the automated testing.  It's just that it looks very much like another bug in the tools used for testing that has been triggered by something in my code.  I can think of things to improve the performance of the code in the presence of invalid input, but I don't think that has anything to do with why the linter is crashing.

**Comment by birkenfeld** [comment]
Without looking further into it, maybe it's a bug in regrxlint handling the `(?i:doctype)` expression, I have never seen it used in pygments. It's also not needed, you can use `(?i)` at the start of the expression.

**Comment by jofish920** [comment]
Thanks.  I'll look into it.  I've never used `regexlint` before.

**Comment by jofish920** [comment]
Thanks again.  The bug was, as suggested, triggered by the use of the "(?i:...)" construct.  I used it out of habit, and will try to avoid doing so in the future.
